### DISCUSSION TITLE
Recognize edge-coordinates of different (always) rectangular objects

### DISCUSSION THREAD

**Discussion by dkk228** [ImageMagick/ImageMagick#2002] [General]
![example](https://user-images.githubusercontent.com/65090992/81495516-0b9e1e80-92b1-11ea-89fc-65e7696acd11.jpg)

Hey there,

the Corona Crisis' advantage for me is to have time for my hobbies. In former times I used to be (to become ;) ) a programmer. I hope you are all fine (at this point). Corona is very serious for me and I hope (and think) that we'll get that together! 

Now my advantage:

I need the coordinates of the rectangles edges. (As shown in the attached picture). The background and the lightning conditions should be able to vary. The shape/contoure is always a rectangle. Varies in ratio, size and the number of rectangles. (One to four should be the average for my expectations)

Is it possible to realize that with IM or am I forced to use openCV? I like IM and I hope that there are efficient solutions for my problem.

Best wishes to all readers!

Yours,

DumbButNice

**Comment by snibgo** [comment]
It is possible, but "the background and the lightning conditions should be able to vary" makes it difficult. IM doesn't have a high-level operation "-detect-background". In your example, the background contains more detail than the objects you want to detect, so a simple edge-detect or similar filter will fail.

**Reply by fmw42** [reply]
If the background is different enough from the foreground, you can try my multicrop or multicrop2 script for bash unix. Alternately, flood fill the background with black and make the foreground white. Then use -connected components to extract the bounding boxes of the objects. Then use the new minimum bounding box to get the rotated bounding box.  What is your IM version and platform/OS?

**Reply by dkk228** [reply]
Thanks and thanks in advance for your mind-time. I'll answer tomorrow. We have Muttertag today (Germany). Quest accepted... But now I have  to relax... 
Have a good night, 
See you soon...

**Comment by dkk228** [comment]
Hey :)
(IM 7.0.10.10 - Win10Pro-1909) ->Future Expansion is to port it to a linux (probably RaspBian)
I tried it with a dark background. If i grayscale and binarize this picture, then I should be able to find the objects borders with an easy loop until (ie) min 300 pixelsInRow notBlack thing, i suppose. Scale down by 10 should fasten the process with less lost of accuracy.

These are my thoughts to solve this problem. But I am a beginner in graphic manipulation, above all if its in commandline :) I am basicly informed about the jpeg compression and formats like bitmap or vector graphics. It's not my goal to get the code line. Here you are... Next... 
I want to understand the backgrounds and how it works. The best case is, if I am able to programm it without any api. (But that is not the goal. I suppose that I do not have enough time in this life to understand everything in graphic manipulations :D )

So long, (it is raining so hard here. Like the eve of deduction :) )

DumbButNice


(Edit: 202005111543)
I realized, that the "min 300pxInRow notBlack" thing is only valid, if the object is in proper orientation. Okay, I like it ;D
The objects are always rectangular (as introduced). Is there a possibility to detect rectangular shapes/contures? Compare it with a vectorized version of a rectangle, something like that?
Are my suggestions sensible?
